vhdx/src/vhdx_metadata.rs

```diff
@@ -0,0 +1,320 @@
+                }
+
+                _ => return Err(Error::InvalidMetadataItem),
+            }
+
+            if (metadata_entry.flag_bits & METADATA_FLAGS_IS_REQUIRED) == 0 {
+                return Err(Error::UnsupportedFlag);
+            }
+            offset += size_of::<MetadataTableEntry>();
+        }
+
+        // Check if all required metadata are present
+        if metadata_presence != METADATA_ALL_PRESENT {
+            return Err(Error::MissingMetadata);
+        }
+        // Check if the virtual disk size is a multiple of the logical sector size
+        if ((metadata_presence & METADATA_LOGICAL_SECTOR_SIZE_PRESENT) != 0)
+            && (disk_spec.virtual_disk_size % disk_spec.logical_sector_size as u64 != 0)
+        {
+            return Err(Error::InvalidBlockSize);
+        }
+
+        disk_spec.sectors_per_block =
+            DiskSpec::sectors_per_block(disk_spec.block_size, disk_spec.logical_sector_size)?;
+
+        disk_spec.chunk_ratio =
+            DiskSpec::chunk_ratio(disk_spec.block_size, disk_spec.logical_sector_size)?;
+
+        disk_spec.total_sectors =
+            disk_spec.virtual_disk_size / disk_spec.logical_sector_size as u64;
+
+        Ok(disk_spec)
+    }
+
+    /// Calculates the number of sectors per block
+    fn sectors_per_block(block_size: u32, logical_sector_size: u32) -> Result<u32> {
+        let sectors_per_block = block_size / logical_sector_size;
+
+        if sectors_per_block & (sectors_per_block - 1) != 0 {
+            return Err(Error::InvalidValue);
+        }
+
+        Ok(sectors_per_block)
+    }
+
+    /// Calculate the chunk ratio
+    fn chunk_ratio(block_size: u32, logical_sector_size: u32) -> Result<u64> {
+        let chunk_ratio = (MAX_SECTORS_PER_BLOCK * logical_sector_size as u64) / block_size as u64;
+
+        if chunk_ratio & (chunk_ratio - 1) != 0 {
+            return Err(Error::InvalidValue);
+        }
+
+        Ok(chunk_ratio)
+    }
+}
+
+#[derive(Default, Debug)]
+struct MetadataTableHeader {
+    signature: u64,
+    _reserved: u16,
+    entry_count: u16,
+    _reserved2: [u8; 20],
+}
+
+impl MetadataTableHeader {
+    pub fn new(buffer: &[u8]) -> Result<MetadataTableHeader> {
+        let metadata_table_header = MetadataTableHeader {
+            signature: LittleEndian::read_u64(&buffer[0..8]),
+            _reserved: Default::default(),
+            entry_count: LittleEndian::read_u16(&buffer[10..12]),
+            ..Default::default()
+        };
+
+        if metadata_table_header.signature != METADATA_SIGN {
+            return Err(Error::InvalidMetadataSign);
+        }
+
+        if metadata_table_header.entry_count > METADATA_MAX_ENTRIES {
+            return Err(Error::InvalidEntryCount);
+        }
+
+        Ok(metadata_table_header)
+    }
+}
+
+#[derive(Default, Debug)]
+pub struct MetadataTableEntry {
+    item_id: u128,
+    offset: u32,
+    length: u32,
+    flag_bits: u32,
+    _reserved: u32,
```

Comment:
Since you read each field in individually (a poor choice imho) you don't need this field.

Reply:
Please provide the alternative you're thinking about so that @fazlamehrab can implement it.

Reply:
To be clear: CH only supports LE systems. So if you keep the reserved fields, mark the struct as `#[repr(packed)]` then you can just read the whole struct from the file in one go.